File: AgoraChat/Code/Controllers/Settings/ACDGeneralViewController.m
//
//  ACDGeneralViewController.m
//  AgoraChat
//
//  Created by liu001 on 2022/3/8.
//  Copyright © 2022 easemob. All rights reserved.
//

#import "ACDGeneralViewController.h"
#import "ACDTitleDetailCell.h"
#import "AgoraGroupPermissionCell.h"
#import "ACDNoDisturbViewController.h"

static NSString *agoraGroupPermissionCellIdentifier = @"AgoraGroupPermissionCell";

@interface ACDGeneralViewController ()
@property (nonatomic,strong) NSString *noDisturbState;

@end

@implementation ACDGeneralViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.navigationItem.leftBarButtonItem = [ACDUtil customLeftButtonItem:@"General" action:@selector(back) actionTarget:self];

    self.noDisturbState = AgoraChatClient.sharedClient.pushManager.pushOptions.silentModeEnabled ?@"ON":@"Off";
    [self.table reloadData];
}


- (void)back {
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark private method
- (void)goNodisturbPage {
    ACDNoDisturbViewController *vc = ACDNoDisturbViewController.new;
    [self.navigationController pushViewController:vc animated:YES];
}

#pragma mark - Table view data source
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 54.0f;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 3;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
   
    
    AgoraGroupPermissionCell *cell = [tableView dequeueReusableCellWithIdentifier:agoraGroupPermissionCellIdentifier];
    if (!cell) {
        cell = [[[NSBundle mainBundle] loadNibNamed:@"AgoraGroupPermissionCell" owner:self options:nil] lastObject];
        cell.permissionSwitch.hidden = NO;
        cell.permissionDescriptionLabel.hidden = YES;
    }
    
    ACDDemoOptions *options = [ACDDemoOptions sharedOptions];

    if(indexPath.row == 0) {
        cell.permissionTitleLabel.text = @"Show Typing";
        [cell.permissionSwitch setOn:options.isChatTyping animated:NO];
        
        cell.switchStateBlock = ^(BOOL isOn) {
            options.isChatTyping = isOn;
            [[ACDDemoOptions sharedOptions] archive];
            [self.table reloadData];
        };
    }else if(indexPath.row == 1) {
        cell.permissionTitleLabel.text = @"Add Group Request";
        BOOL autoAcceptGroupRequest = AgoraChatClient.sharedClient.options.autoAcceptGroupInvitation;
        [cell.permissionSwitch setOn:autoAcceptGroupRequest animated:NO];
        cell.switchStateBlock = ^(BOOL isOn) {
            [AgoraChatClient.sharedClient.options setAutoAcceptGroupInvitation:isOn];
            options.isAutoAcceptGroupInvitation = isOn;
            [options archive];
            [self.table reloadData];
        };
    }else {
        cell.permissionTitleLabel.text = @"Delete the Chat after Leaving Group";
        BOOL deleteMessagesOnLeaveGroup = AgoraChatClient.sharedClient.options.deleteMessagesOnLeaveGroup;
        [cell.permissionSwitch setOn:deleteMessagesOnLeaveGroup animated:NO];
        cell.switchStateBlock = ^(BOOL isOn) {
            [AgoraChatClient.sharedClient.options setDeleteMessagesOnLeaveGroup:isOn];
            [self.table reloadData];
        };
    }
    
    return cell;
}

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Settings/ACDGeneralViewController.swift
//
//  ACDGeneralViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/13.
//

import UIKit

class ACDGeneralViewController: UITableViewController {
    
    @IBOutlet weak var showTypingSwitch: UISwitch!
    @IBOutlet weak var autoAcceptGroupInviteSwitch: UISwitch!
    @IBOutlet weak var deleteChatSwitch: UISwitch!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.navigationItem.leftBarButtonItem = ACDUtil.customLeftButtonItem(title: "General".localized, action: #selector(back), target: self)
        
        self.showTypingSwitch.isOn = ACDDemoOptions.shared.isChatTyping
        self.autoAcceptGroupInviteSwitch.isOn = ACDDemoOptions.shared.isAutoAcceptGroupInvitation
        self.deleteChatSwitch.isOn = ACDDemoOptions.shared.deleteMessagesOnLeaveGroup
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setNavigationBarHidden(false, animated: true)
    }
    
    @objc private func back() {
        self.navigationController?.popViewController(animated: true)
    }

    @IBAction func switchValueChange(_ sender: UISwitch) {
        if sender == self.showTypingSwitch {
            ACDDemoOptions.shared.isChatTyping = sender.isOn
        } else if sender == self.autoAcceptGroupInviteSwitch {
            AgoraChatClient.shared.options.autoAcceptGroupInvitation = sender.isOn
            ACDDemoOptions.shared.isAutoAcceptGroupInvitation = sender.isOn
        } else if sender == self.deleteChatSwitch {
            AgoraChatClient.shared.options.deleteMessagesOnLeaveGroup = sender.isOn
            ACDDemoOptions.shared.deleteMessagesOnLeaveGroup = sender.isOn
        }
        ACDDemoOptions.shared.archive()
    }
    
    override func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 0.1
    }
    
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 0.1
    }
}
